core/store/src/cold_storage.rs

```diff
@@ -0,0 +1,255 @@
+use crate::columns::DBKeyType;
+use crate::{DBCol, DBTransaction, Database, Store};
+
+use borsh::BorshDeserialize;
+use near_primitives::types::BlockHeight;
+use std::collections::HashMap;
+use std::io;
+use std::sync::Arc;
+use strum::IntoEnumIterator;
+
+type StoreKey = Vec<u8>;
+type StoreValue = Option<Vec<u8>>;
+type StoreCache = HashMap<(DBCol, StoreKey), StoreValue>;
+
+struct StoreWithCache<'a> {
+    store: &'a Store,
+    cache: StoreCache,
+}
+
+pub fn update_cold_db(
+    cold_db: &Arc<dyn Database>,
```

Comment:
No need for `Arc`.  `cold_db: &dyn Database` is sufficient.

Also, this may be personal preference, but I find `foo(destination, source)` more natural.  `fn update_cold_db(hot_store, cold_db, height)`.  Up to you.